shardedSearcher - implement functionality for filetombstones

ggilmore opened the pull request:
This PR extracts the searcher functionality from the incremental work in https://github.com/sourcegraph/zoekt/pull/257 so that it can be merged separately. 

--- 

This PR does a few things:
- Adds a new field, `FileTombstones`  to the zoekt.Repository type. This field represents a set of file paths that should be ignored (across all branches) from this shard's search results. 
	- Right now, the set of FileTombstones is implemented as a hashset of file names. This isn't the best performance-wise, but it seemed like the most straightforward solution at the time. We can do better by keeping a sorted list of document IDs instead, but I will leave it for y'all to comment on whether or not we need it now, or closer to when we want to enable this feature.
- ShardedSearcher knows how to consult FileTombstones when calculating search results, and skipping over files that match one of the tombstones. 
	- A test is added for this functionality
- Adds a new SearchOption ,`EnableFileTombstones`, that feature flags the behavior above. Right now it isn't possible to programmatically enable/disable the feature (I figured it wasn't worth it since this PR alone doesn't implement a change that actually sets the filetombstones field). I figure that we can revisit this once the other parts land. 

Note that this PR doesn't tell builder to properly set filetombstones - this is only the searcher portion of the functioanlity. The builder logic will live in https://github.com/sourcegraph/zoekt/pull/274

ggilmore committed:
shardedSearcher - implement functionality for filetombstones

ggilmore commented:
cc @sourcegraph/search-core There is a test failure here. I think it's just complaining about the new field that I added. Is this benign (since we never modify existing shards, only the `.meta` file)? I can proceed by either :
1.  trying to regenerate the test data 
1.  adding `omitempty` to the `FileTombstones` field in the zoekt.Repository definition.

LMK
![Screen Shot 2022-02-14 at 1 28 13 PM](https://user-images.githubusercontent.com/9022011/153949785-c9f51423-5116-44ed-abc4-698abc0e380c.png)

ggilmore commented on shards/shards_test.go:
```diff
@@ -958,6 +959,175 @@ func TestAtomCountScore(t *testing.T) {
 	}
 }
 
+func TestSearchFileTombstones(t *testing.T) {
```
I'd also appreciate feedback on the tests. One thing that strikes me as odd is that asserting on the number of line matches seems brittle / maybe asserting too much about internals.

keegancsmith replied:
two comments:

- you can have a single corpus of documents / repository. Then each test case only alternates what you care about `FileTombstones`. Smaller test descriptions lead to easier grokking of what the test is doing.
- Yeah line match counts doesn't give me confidence. I think maybe you should assert on the set of filenames returned.

ggilmore commented on eval.go:
```diff
@@ -237,14 +237,28 @@ nextFileMatch:
 
+			if opts.EvaluateFileTombstones {
+
+				if repoMetadata.FileTombstones != nil {
```
Having to manage a possibly `nil` FileTombstones map also strikes me as awkward, and maybe we should move `FileTombstones` to the sorted list of document IDs approach sooner rather than later. (The other ways that I can think about this are to have nil guards everywhere, or to mandate the use of a factory `zoket.NewRepository()` that takes care of initializing the field for you - neither of these seem great).

keegancsmith replied:
you can do lookups on nil maps. This works and `ok` is false

```go
var m map[string]struct{}
m = nil
_, ok := m["foo"]
```

that should simplify this control flow a bit.

Also you can skip map lookups if `len(...)` is zero. Benefit of this is you skip the more expensive fileName creation work.

keegancsmith commented on api.go:
```diff
@@ -307,6 +307,10 @@ type Repository struct {
+
+	// FileTombstones is a set of file paths that should be ignored across all branches
+	// in this shard.
+	FileTombstones map[string]struct{}
```
maybe add a json `omitempty` so this field doesn't appear unless it has values?

keegancsmith commented on api.go:
```diff
@@ -541,6 +545,12 @@ type SearchOptions struct {
+	// field when evaluating a search request.
+	//
+	// Note: This feature flag is off by default.
+	EvaluateFileTombstones bool
```
I'm unsure if we need to feature flag, given this field won't be set unless we set it. IE once a shard contains tombstones, lets use them!

keegancsmith replied:
Yeah I still think we should not have the feature flag. If tombstones are present we should read them.

keegancsmith commented on eval.go:
```diff
@@ -237,14 +237,28 @@ nextFileMatch:
+			if opts.EvaluateFileTombstones {
+
+				if repoMetadata.FileTombstones != nil {
+					fileName := string(d.fileName(nextDoc))
```
ouch, this is a lot more work than what we did before. `d.fileName` takes work to lookup if I am not mistaken. and we are allocating a string on the heap. I'm fairly convinced before this is ready to ship for reals, we need to switch to a document ID based representation. However, we can ship this PR before that for testing purposes / minimize the amount of work you are doing.

keegancsmith commented on eval.go:
```diff
@@ -237,14 +237,28 @@ nextFileMatch:
 
 		for ; nextDoc < docCount; nextDoc++ {
+			repoID := d.repos[nextDoc]
+			repoMetadata := d.repoMetaData[repoID]
```
a bit of a footgun, but you just did a copy of the repository field since the slice is to values not pointers. Previously we didn't do this. To keep the readability and avoid the copy you can do the following
```suggestion
			repoMetadata := &d.repoMetaData[repoID]
```

keegancsmith reviewed (commented):
Interested in @stefanhengl's thoughts on landing this now, even though we probably wanna switch to document IDs. I think it makes sense since the code path will only be active for shards with tombstones, so gives us smaller chunks to review.

keegancsmith commented:
> 1. trying to regenerate the test data
> 2. adding `omitempty` to the `FileTombstones` field in the zoekt.Repository definition.

I like option 2.

ggilmore committed:
add json omitempty to filetombstones definition

ggilmore committed:
re-write tests according to feedback

(drops line matching, and instead compares the set of returned documents)
(matching only on the filename is insufficient since there are test where a file is tombstoned in one shard, but present in another)

ggilmore committed:
Update eval.go

Co-authored-by: Keegan Carruthers-Smith <keegan.csmith@gmail.com>

ggilmore committed:
remove extraneous nil check

ggilmore committed:
skip checking for filetombstones if length of the set is <= 0

ggilmore commented:
PTAL, I responded to most of the feedback and re-wrote the tests. cc @stefanhengl too. 

I am okay with removing the feature flag for this PR since nobody should be setting it at the moment in production.

keegancsmith commented on eval.go:
```diff
@@ -237,14 +237,24 @@ nextFileMatch:
+			repoID := d.repos[nextDoc]
+			repoMetadata := &d.repoMetaData[repoID]
+
 			// Skip tombstoned docs
```
minor can you update this docstring to say we are skipping repos that are tombstoned

keegancsmith commented on eval.go:
```diff
@@ -237,14 +237,24 @@ nextFileMatch:
 				continue
 			}
 
+			if opts.EvaluateFileTombstones && len(repoMetadata.FileTombstones) > 0 {
```
minor: add a comment that this is for document tombstones. Maybe leave in a TODO saying this design is doing a lot of extra work due to needing to fetch filenames and allocate strings.

ggilmore replied:
``` 
TODO: This FileTombstones implementation (looking up by filenames) creates a lot of small allocations
			// (string filenames) and can have poor cache performance. This should be addressed before we officially
			// roll this out.
```

keegancsmith commented on shards/shards_test.go:
```diff
@@ -958,6 +958,167 @@ func TestAtomCountScore(t *testing.T) {
+				receivedDocuments = append(receivedDocuments, simpleDocument{
+					Name:     f.FileName,
+					Branches: f.Branches,
+					Content:  f.Content,
```
why include content? In fact, why include branches? Our implementation will exclude filenames no matter the branch. So we should just need to test on the set of filenames. This should also simplify the comparison code since you have is a `[]string` that you can sort (or just a `map[string]struct{}`)

ggilmore replied:
There is a test that reads from two shards and asserts the tombstoned foo.go doesn't show up in the results, but the results from the non-tombstoned foo.go in the other shard do. In this case, you can't assert on just file names - you have to look at the contents

ggilmore replied:
However, I think you are correct about the branches. Each file has unique content, so it's redundant to assert on the branches as well. I'll remove them from the comparison.

ggilmore committed:
update tombstoned comment

ggilmore committed:
remove evaluatefiletombstones feature flag

ggilmore committed:
add comment about poor performance

ggilmore committed:
remove branches from test comparison

each file has unique content, so asserting on branches is redundant
